### PR TITLE
fix #7187

### LINKED ISSUE #7187: sequtils.toSeq produces to sequence from the iterator twice if compiles(iter.len) == true
In the doc comment for sequtils.toSeq we find this example:
```nim
let
  numeric = @[1, 2, 3, 4, 5, 6, 7, 8, 9]
  odd_numbers = toSeq(filter(numeric) do (x: int) -> bool:
         if x mod 2 == 1:
           result = true)
```
If you run this, the filtered sequence is produced twice, as can be seen [here on the nim playground](https://play.nim-lang.org/?gist=7d4a355b9d440a1fbbee810a7edbf3a0) with my instrumented copies of `filter` and `toSeq`. I think its the same problem that was fixed eg. in mapIt by copying the iterable to a local variable and therefore forcing the iteration to take place.

My proposed fix looks like this:
```nim
template toSeqFixed*(iter: untyped): untyped =
  when compiles(iter.len):
    iter
  else:
    var result: seq[type(iter)] = @[]
    for x in iter:
      result.add(x)
    result
```
but, as usual, i am not sure if this fix is correct...

### PR DESCRIPTION
**Pull request by hlaaftana**
A bit overdue

Fixes #7187

### CODE REVIEW

File: lib/pure/collections/sequtils.nim [hunk 1]
```diff
@@ -497,9 +497,10 @@ template toSeq*(iter: untyped): untyped =
   ##   assert odd_numbers == @[1, 3, 5, 7, 9]
 
   when compiles(iter.len):
+    let s = iter
```
Comment (Varriount) [inline lib/pure/collections/sequtils.nim]: My only concern here is that if `iter` is an object (say, a table) the assignment will copy the entire structure, which may be quite expensive.

**Comment by hlaaftana** [comment]
I tried to squash 2 commits and got this

image

**Comment by Yardanico** [comment]
@hlaaftana you don't need to squash commits by yourself! The person who merges your PR can select an option to squash all commits automatically

**Comment by timotheecour** [comment]
=> proposed a different fix here: https://github.com/nim-lang/Nim/pull/8586

**Pull request closed without merging**